CI: enable cpp and c backends

It looks like after we migrated to github we did not enable the cpp tests and they don't seem to run:
```console
$ ./run_tests.py -j1 -b cpp
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-llvm
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-llvm2
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-llvm_rtlib
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-cpp
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-x86
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-wasm
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-gfortran
+ rm -rf /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-llvmImplicit
+ mkdir /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-cpp
+ FC=lfortran cmake -DLFORTRAN_BACKEND=cpp -DCURRENT_BINARY_DIR=/Users/ondrej/repos/lfortran/lfortran/integration_tests/test-cpp -S /Users/ondrej/repos/lfortran/lfortran/integration_tests -B /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-cpp
-- The C compiler identification is AppleClang 13.0.0.13000029
-- The Fortran compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - failed
-- Check for working Fortran compiler: /Users/ondrej/bin/lfortran
-- Check for working Fortran compiler: /Users/ondrej/bin/lfortran - works
-- Checking whether /Users/ondrej/bin/lfortran supports Fortran 90
-- Checking whether /Users/ondrej/bin/lfortran supports Fortran 90 - yes


Configuration results
---------------------
Fortran compiler: /Users/ondrej/bin/lfortran
C compiler      : /Library/Developer/CommandLineTools/usr/bin/cc
Build type: Debug
Fortran compiler flags: 
C compiler flags      : -g
Installation prefix: /usr/local
LFORTRAN_BACKEND: cpp
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ondrej/repos/lfortran/lfortran/integration_tests/test-cpp
+ make -j1
Scanning dependencies of target program_cmake_01
[  1%] Building Fortran object CMakeFiles/program_cmake_01.dir/program_cmake_01.f90.o
CMakeFiles/program_cmake_01.dir/program_cmake_01.f90.o.tmp.cpp:8:10: fatal error: 'lfortran_intrinsics.h' file not found
#include <lfortran_intrinsics.h>
         ^~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
The command 'g++ -std=c++17 -I/Users/ondrej/ext/include -I/Users/ondrej/repos/lfortran/lfortran/src/bin/../runtime/impure -o CMakeFiles/program_cmake_01.dir/program_cmake_01.f90.o -c CMakeFiles/program_cmake_01.dir/program_cmake_01.f90.o.tmp.cpp' failed.
make[2]: *** [CMakeFiles/program_cmake_01.dir/program_cmake_01.f90.o] Error 11
make[1]: *** [CMakeFiles/program_cmake_01.dir/all] Error 2
make: *** [all] Error 2
Command failed.
```

But it should not be difficult to fix.

--- a/.github/workflows/CI.yml
+++ b/.github/workflows/CI.yml
@@ -597,8 +597,8 @@ jobs:
             cd integration_tests
             ./run_tests.py -b wasm
 
-  test_cpp:
-    name: Test C++ backend
+  test_c_cpp:
+    name: Test C/C++ backend
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
@@ -649,7 +649,7 @@ jobs:
             cd ../../..
 
             cd integration_tests
-            ./run_tests.py -j1 -b cpp
+            ./run_tests.py -j1 -b cpp c
 
   upload_tarball:
     name: Upload Tarball

--- a/.gitignore
+++ b/.gitignore
@@ -93,6 +93,8 @@ input.txt
 input
 src/libasr/libasr.a.*
 *.html
+*.o.tmp.c
+*.o.tmp.cpp
 
 ### https://raw.github.com/github/gitignore/218a941be92679ce67d0484547e3e142b2f5f6f0/Global/macOS.gitignore
 

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -73,7 +73,7 @@ macro(RUN_UTIL RUN_FAIL RUN_NAME RUN_FILE_NAME RUN_INCLUDE_PATH RUN_LABELS RUN_E
 			set(include_path_with_flag "")
 		endif()
 
-        if ((LFORTRAN_BACKEND STREQUAL "cpp") OR (LFORTRAN_BACKEND STREQUAL "x86"))
+        if ((LFORTRAN_BACKEND STREQUAL "cpp") OR (LFORTRAN_BACKEND STREQUAL "x86") OR (LFORTRAN_BACKEND STREQUAL "c"))
             add_executable(${name} ${file_name}.f90 ${extra_files})
             target_compile_options(${name} PUBLIC ${include_path_with_flag} ${extra_args} --backend=${LFORTRAN_BACKEND})
             target_link_options(${name} PUBLIC --backend=${LFORTRAN_BACKEND})
@@ -185,7 +185,7 @@ RUN(NAME include_03 LABELS gfortran llvm wasm INCLUDE_PATH include_03)
 RUN(NAME cond_01 LABELS gfortran llvm cpp x86 wasm)
 RUN(NAME cond_02 LABELS gfortran llvm wasm)
 
-RUN(NAME expr_01 FAIL LABELS gfortran llvm cpp x86 wasm)
+RUN(NAME expr_01 FAIL LABELS gfortran llvm cpp x86 wasm c)
 RUN(NAME expr_02 LABELS gfortran llvm cpp x86 wasm)
 RUN(NAME expr_03 LABELS gfortran llvm cpp x86 wasm)
 RUN(NAME expr_04 LABELS gfortran llvm cpp wasm)

--- a/integration_tests/run_tests.py
+++ b/integration_tests/run_tests.py
@@ -6,7 +6,8 @@ import os
 
 # Initialization
 NO_OF_THREADS = 8 # default no of threads is 8
-SUPPORTED_BACKENDS = ['llvm', 'llvm2', 'llvm_rtlib', 'cpp', 'x86', 'wasm', 'gfortran', 'llvmImplicit']
+SUPPORTED_BACKENDS = ['llvm', 'llvm2', 'llvm_rtlib', 'c',
+                      'cpp', 'x86', 'wasm', 'gfortran', 'llvmImplicit']
 BASE_DIR = os.path.dirname(os.path.realpath(__file__))
 LFORTRAN_PATH = f"{BASE_DIR}/../src/bin:$PATH"
 

--- a/src/bin/lfortran.cpp
+++ b/src/bin/lfortran.cpp
@@ -60,7 +60,7 @@ using LCompilers::endswith;
 using LCompilers::CompilerOptions;
 
 enum Backend {
-    llvm, cpp, x86, wasm
+    llvm, c, cpp, x86, wasm
 };
 
 std::string read_file(const std::string &filename)
@@ -1266,6 +1266,107 @@ int compile_to_object_file_cpp(const std::string &infile,
     return 0;
 }
 
+int compile_to_object_file_c(const std::string &infile,
+        const std::string &outfile,
+        bool assembly, const std::string &rtlib_header_dir,
+        CompilerOptions &compiler_options)
+{
+    std::string input = read_file(infile);
+
+    LCompilers::FortranEvaluator fe(compiler_options);
+    LCompilers::ASR::TranslationUnit_t* asr;
+
+    // Src -> AST -> ASR
+    LCompilers::LocationManager lm;
+    {
+        LCompilers::LocationManager::FileLocations fl;
+        fl.in_filename = infile;
+        lm.files.push_back(fl);
+        lm.file_ends.push_back(input.size());
+    }
+    LCompilers::diag::Diagnostics diagnostics;
+    LCompilers::Result<LCompilers::ASR::TranslationUnit_t*>
+        result = fe.get_asr2(input, lm, diagnostics);
+    std::cerr << diagnostics.render(lm, compiler_options);
+    if (result.ok) {
+        asr = result.result;
+    } else {
+        LCOMPILERS_ASSERT(diagnostics.has_error())
+        return 1;
+    }
+
+    // Save .mod files
+    {
+        int err = save_mod_files(*asr, compiler_options);
+        if (err) return err;
+    }
+
+    if (!LCompilers::ASRUtils::main_program_present(*asr)) {
+        // Create an empty object file (things will be actually
+        // compiled and linked when the main program is present):
+        if (compiler_options.platform == LCompilers::Platform::Windows) {
+            {
+                std::ofstream out;
+                out.open(outfile);
+                out << " ";
+            }
+        } else {
+            std::string outfile_empty = outfile + ".empty.c";
+            {
+                std::ofstream out;
+                out.open(outfile_empty);
+                out << " ";
+            }
+	    std::string CC = "cc";
+            char *env_CC = std::getenv("LFORTRAN_CC");
+            if (env_CC) CC = env_CC;
+            std::string cmd = CC + " -c '" + outfile_empty + "' -o '" + outfile + "'";
+            int err = system(cmd.c_str());
+            if (err) {
+                std::cout << "The command '" + cmd + "' failed." << std::endl;
+                return 11;
+            }
+        }
+        return 0;
+    }
+
+    // ASR -> C
+    std::string src;
+    diagnostics.diagnostics.clear();
+    LCompilers::Result<std::string> res
+        = fe.get_c2(*asr, diagnostics, 1);
+    std::cerr << diagnostics.render(lm, compiler_options);
+    if (res.ok) {
+        src = res.result;
+    } else {
+        LCOMPILERS_ASSERT(diagnostics.has_error())
+        return 5;
+    }
+
+    // C -> Machine code (saves to an object file)
+    if (assembly) {
+        throw LCompilers::LCompilersException("Not implemented");
+    } else {
+        std::string cfile = outfile + ".tmp.c";
+        {
+            std::ofstream out;
+            out.open(cfile);
+            out << src;
+        }
+
+        std::string CXX = "gcc";
+        std::string options = " -I" + rtlib_header_dir;
+        std::string cmd = CXX + " " + options + " -o " + outfile + " -c " + cfile;
+        int err = system(cmd.c_str());
+        if (err) {
+            std::cout << "The command '" + cmd + "' failed." << std::endl;
+            return 11;
+        }
+    }
+
+    return 0;
+}
+
 // infile is an object file
 // outfile will become the executable
 int link_executable(const std::vector<std::string> &infiles,
@@ -1404,6 +1505,22 @@ int link_executable(const std::vector<std::string> &infiles,
                     }
                 }
             }
+        }
+        return 0;
+    } else if (backend == Backend::c) {
+        std::string CXX = "gcc";
+        std::string cmd = CXX + " -o " + outfile + " ";
+        std::string base_path = "\"" + runtime_library_dir + "\"";
+        std::string runtime_lib = "lfortran_runtime";
+        for (auto &s : infiles) {
+            cmd += s + " ";
+        }
+        cmd += " -L" + base_path
+            + " -Wl,-rpath," + base_path + " -l" + runtime_lib + " -lm";
+        int err = system(cmd.c_str());
+        if (err) {
+            std::cout << "The command '" + cmd + "' failed." << std::endl;
+            return 10;
         }
         return 0;
     } else if (backend == Backend::cpp) {
@@ -1820,6 +1937,8 @@ int main(int argc, char *argv[])
 
         if (arg_backend == "llvm") {
             backend = Backend::llvm;
+        } else if (arg_backend == "c") {
+            backend = Backend::c;
         } else if (arg_backend == "cpp") {
             backend = Backend::cpp;
         } else if (arg_backend == "x86") {
@@ -1947,6 +2066,9 @@ int main(int argc, char *argv[])
                 std::cerr << "The -c option requires the LLVM backend to be enabled. Recompile with `WITH_LLVM=yes`." << std::endl;
                 return 1;
 #endif
+            } else if (backend == Backend::c) {
+                return compile_to_object_file_c(arg_file, outfile, false,
+                        rtlib_c_header_dir, compiler_options);
             } else if (backend == Backend::cpp) {
                 return compile_to_object_file_cpp(arg_file, outfile, false,
                         true, rtlib_c_header_dir, compiler_options);
@@ -1981,6 +2103,9 @@ int main(int argc, char *argv[])
             } else if (backend == Backend::cpp) {
                 err = compile_to_object_file_cpp(arg_file, tmp_o, false,
                         true, rtlib_header_dir, compiler_options);
+            } else if (backend == Backend::c) {
+                err = compile_to_object_file_c(arg_file, tmp_o,
+                        false, rtlib_c_header_dir, compiler_options);
             } else {
                 throw LCompilers::LCompilersException("Backend not supported");
             }

--- a/src/lfortran/fortran_evaluator.cpp
+++ b/src/lfortran/fortran_evaluator.cpp
@@ -462,7 +462,7 @@ Result<std::string> FortranEvaluator::get_c(const std::string &code,
 Result<std::string> FortranEvaluator::get_c2(ASR::TranslationUnit_t &asr,
         diag::Diagnostics &diagnostics, int64_t default_lower_bound)
 {
-    // ASR -> C++
+    // ASR -> C
     return asr_to_c(al, asr, diagnostics, compiler_options,
                     default_lower_bound);
 }

